### ISSUE TITLE
Does anyone recommend the way packaging files that used glimmer-dsl-tk to executables?

### ISSUE THREAD

**Issue by Vucius** [AndyObtiva/glimmer-dsl-tk#12]
I try to use `ocra` to packaging the files on `win10 OS`.It worked on the files that used `glimmer-dsl-libui`, but can't on the that used `glimmer-dsl-tk`.
```
/lib/ruby/3.0.0/rubygems/core_ext/kernel_require.rb>:85:in `require': cannot load such file -- glimmer-dsl-tk (LoadError)
```
I did the packaging-work on two pc, they all reported the same error.

So, I come here to ask for the succed way to packaging the files that used `glimmer-dsl-tk`.

**Comment by AndyObtiva** [comment]
Tk requires installing ActiveTcl (mentioned in prerequisites: https://github.com/AndyObtiva/glimmer-dsl-tk#pre-requisites) , which lives outside of Ruby projects. So, I would expect that packaging a Ruby Tk project with OCRA would not work because it would be missing ActiveTcl dependencies. I don’t have much experience with OCRA though, and haven’t had much chance to package native executables from Glimmer DSL for Tk apps yet (it’s a far future task), so it’s probably better to ask your questions at the OCRA project concerning Ruby Tk apps in general (not just Glimmer DSL for Tk, which only adds pure Ruby extra code on top of Tk). Maybe you just have to pass extra options to include ActiveTcl dependencies.

**Comment by kojix2** [comment]
Hi @Vucius, @AndyObtiva 
I have seen [the glimmer-dsl-libui issue](https://github.com/AndyObtiva/glimmer-dsl-libui/issues/35) and would like to make a few comments.

I have never packaged glimmer-dsl-tk, but tk could be packaged around Ruby 2.6, or 7 years ago, with the following command

```
ocra tk.rb ^
  C:\Ruby26-x64\msys64\usr\bin\msys-2.0.dll ^
  C:\Ruby26-x64\msys64\mingw64\lib\tcl8.6 ^
  C:\Ruby26-x64\msys64\mingw64\lib\tk8.6 ^
  --no-autoload --add-all-core
```

I remember that the `--no-lzma` option made the startup much faster. tk started at a practical speed. green shoes (which depends on gtk) could have been packaged in a similar way, but it took tens of seconds to start up and I didn't feel it was very practical.

But that was seven years ago. The situation will be different now. In the ocra issue thread, someone offered the following command

https://github.com/larsch/ocra/issues/131

```
ocra test.rb ^
  --dll ruby_builtin_dlls\libssp-0.dll ^
  --dll ruby_builtin_dlls\libssl-1_1-x64.dll ^
  --dll ruby_builtin_dlls\libcrypto-1_1-x64.dll ^
  --gem-file ^
  C:\Ruby26-x64\ssl C:\Ruby26-x64\msys64\usr\bin\msys-2.0.dll  ^
  C:\Ruby26-x64\msys64\mingw64\bin\tk86.dll ^
  C:\Ruby26-x64\msys64\mingw64\lib\tcl8.6 ^
  C:\Ruby26-x64\msys64\mingw64\lib\tk8.6 ^
  --no-autoload ^
  --add-all-core ^
  --no-lzma ^
  --gem-all ^
  --output .\release\test.exe ^
  --debug-extract ^
  --chdir-first
```

> ruby_builtin_dlls\libssl-1_1-x64.dll
> ruby_builtin_dlls\libcrypto-1_1-x64.dll
> C:\Ruby26-x64\ssl
> these should be specific to my app.

It may be possible to do this today in a similar way as before.

Good luck!

**Comment by AndyObtiva** [comment]
Thank you @kojix2 for the tip!

@Vucius if you succeed at packaging Glimmer DSL for Tk apps, please share the instructions that worked, and we could close this issue. 

Also, if you do manage to get packaging working, you could contribute instructions to README.md as a Pull Request, and thus get added to the list of people who added commits to the project. 

Cheers.

**Comment by Vucius** [comment]
> Hi @Vucius, @AndyObtiva I have seen [the glimmer-dsl-libui issue](https://github.com/AndyObtiva/glimmer-dsl-libui/issues/35) and would like to make a few comments.
> 
> I have never packaged glimmer-dsl-tk, but tk could be packaged around Ruby 2.6, or 7 years ago, with the following command
> 
> ```
> ocra tk.rb ^
>   C:\Ruby26-x64\msys64\usr\bin\msys-2.0.dll ^
>   C:\Ruby26-x64\msys64\mingw64\lib\tcl8.6 ^
>   C:\Ruby26-x64\msys64\mingw64\lib\tk8.6 ^
>   --no-autoload --add-all-core
> ```
> 
> I remember that the `--no-lzma` option made the startup much faster. tk started at a practical speed. green shoes (which depends on gtk) could have been packaged in a similar way, but it took tens of seconds to start up and I didn't feel it was very practical.
> 
> But that was seven years ago. The situation will be different now. In the ocra issue thread, someone offered the following command
> 
> [larsch/ocra#131](https://github.com/larsch/ocra/issues/131)
> 
> ```
> ocra test.rb ^
>   --dll ruby_builtin_dlls\libssp-0.dll ^
>   --dll ruby_builtin_dlls\libssl-1_1-x64.dll ^
>   --dll ruby_builtin_dlls\libcrypto-1_1-x64.dll ^
>   --gem-file ^
>   C:\Ruby26-x64\ssl C:\Ruby26-x64\msys64\usr\bin\msys-2.0.dll  ^
>   C:\Ruby26-x64\msys64\mingw64\bin\tk86.dll ^
>   C:\Ruby26-x64\msys64\mingw64\lib\tcl8.6 ^
>   C:\Ruby26-x64\msys64\mingw64\lib\tk8.6 ^
>   --no-autoload ^
>   --add-all-core ^
>   --no-lzma ^
>   --gem-all ^
>   --output .\release\test.exe ^
>   --debug-extract ^
>   --chdir-first
> ```
> 
> > ruby_builtin_dlls\libssl-1_1-x64.dll
> > ruby_builtin_dlls\libcrypto-1_1-x64.dll
> > C:\Ruby26-x64\ssl
> > these should be specific to my app.
> 
> It may be possible to do this today in a similar way as before.
> 
> Good luck!

thx.  I did packaging .rb to .exe successfully with `ocra`, even that .exe couldn't work correctly when `ocra` need to find 'tk'...
anyway,thx for ur tips.

**Comment by Vucius** [comment]
> Reference in new

When I used ruby-2.6.5, `ocra` would packaging `.rb` that use glimmer-dsl-tk:
```
ocra test.rb --dll ruby_builtin_dlls/libssp-0.dll --dll ruby_builtin_dlls/libgmp-10.dll /c/Ruby26-x64/msys64/usr/bin/msys-2.0.dll /c/Ruby26-x64/bin/tcl86t.dll /c/Ruby26-x64/bin/tk86t.dll --no-autoload
```
 `.exe` will work correctly.
Worsely, `ocra` will be stucked after user inputed command some option.Strictnessly, I think `ocra` exists some bug that it couldn't load `tk` totally.yes, just like u mentioned,the problem isn't `glimmer-dsl-tk` .So I closed this issue, albeit I want to know the other way to packging `.rb`.
I also tried use `ruby-packer` to packaging `.exe`. Unfortunately, `ruby-packer` lacks some option to load `tk` sources.
**I will be appreciated if someone share other way to packging `.rb`.**
